controller/src/util/instance_action.rs

```diff
@@ -53,107 +91,30 @@ pub enum InstanceAction {
-) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
-    trace!("internal_do_instance_watch - enter");
-    let resource = Api::<Instance>::all(kube_interface.get_kube_client());
-    let watcher = watcher(resource, Config::default()).default_backoff();
-    let mut informer = watcher.boxed();
-    let mut first_event = true;
-    // Currently, this does not handle None except to break the loop.
-    loop {
-        let event = match informer.try_next().await {
-            Err(e) => {
-                error!("Error during watch: {}", e);
-                continue;
-            }
-            Ok(None) => break,
-            Ok(Some(event)) => event,
-        };
-        // Aquire lock to ensure cleanup_instance_and_configuration_svcs and the
-        // inner loop handle_instance call in internal_do_instance_watch
-        // cannot execute at the same time.
-        let _lock = synchronization.lock().await;
-        trace!("internal_do_instance_watch - aquired sync lock");
-        handle_instance(event, kube_interface, &mut first_event).await?;
-    }
-    Ok(())
-}
-
-/// This takes an event off the Instance stream and delegates it to the
-/// correct function based on the event type.
-async fn handle_instance(
-    event: Event<Instance>,
-    kube_interface: &impl KubeInterface,
-    first_event: &mut bool,
-) -> anyhow::Result<()> {
-    trace!("handle_instance - enter");
+async fn reconcile_inner(event: Event<Instance>, ctx: Arc<ControllerContext>) -> Result<Action> {
     match event {
-        Event::Applied(instance) => {
-            info!(
-                "handle_instance - added or modified Akri Instance {:?}: {:?}",
-                instance.metadata.name, instance.spec
-            );
-            // TODO: consider renaming `InstanceAction::Add` to `InstanceAction::AddOrUpdate`
-            // to reflect that this could also be an Update event. Or as we do more specific
-            // inspection in future, delineation may be useful.
-            handle_instance_change(&instance, &InstanceAction::Add, kube_interface).await?;
+        Event::Apply(instance) => {
+            handle_instance_change(&instance, &InstanceAction::Add, ctx.clone()).await
         }
-        Event::Deleted(instance) => {
-            info!(
-                "handle_instance - deleted Akri Instance {:?}: {:?}",
-                instance.metadata.name, instance.spec
-            );
-            handle_instance_change(&instance, &InstanceAction::Remove, kube_interface).await?;
-        }
-        Event::Restarted(_instances) => {
-            if *first_event {
-                info!("handle_instance - watcher started");
-            } else {
-                return Err(anyhow::anyhow!(
-                    "Instance watcher restarted - throwing error to restart controller"
-                ));
-            }
+        Event::Cleanup(instance) => {
+            handle_instance_change(&instance, &InstanceAction::Remove, ctx.clone()).await
```

Comment:
Just thinking out loud here, but since the Pods/Jobs are owned by the Instance, deleting it should delete all owned pods/jobs leaving nothing to do for the controller, is it really needed for us to do this part then ?
This would save us from adding a finalizer to the Instance as well.

Reply:
Good call. I agree that we can do nothing on the cleanup case